### PR TITLE
UCT/IB/RC: tag matching infrastructure

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe  please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1114/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3047/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1122/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3056/ for details (Mellanox internal link).

File: src/uct/ib/rc/verbs/rc_verbs_iface.c [hunk 1]
```diff
@@ -96,13 +108,286 @@ void uct_rc_verbs_iface_progress(void *arg)
...
+}
+
+static void uct_rc_verbs_iface_init_inl_wrs(uct_rc_verbs_iface_t *iface)
+{
+    unsigned num_am_sge;
+
+#if HAVE_IBV_HW_TM
+    if (UCT_RC_VERBS_TM_ENABLED(iface)) {
+        num_am_sge                            = 3;
+        iface->verbs_common.inl_sge[0].addr   = (uintptr_t)iface->tm.notag_hdr;
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_iface.c]: if used for inline, might be better to unite am_hdr with notag_hdr

File: src/uct/ib/base/ib_device.h [hunk 2]
```diff
@@ -12,6 +12,7 @@
 #include <uct/api/uct.h>
 #include <ucs/stats/stats.h>
 #include <ucs/debug/log.h>
+#include <endian.h>
```
Comment (shamisp) [inline src/uct/ib/base/ib_device.h]: why ?

File: src/uct/ib/rc/base/rc_iface.c [hunk 3]
```diff
@@ -401,18 +401,22 @@ UCS_CLASS_INIT_FUNC(uct_rc_iface_t, uct_rc_iface_ops_t *ops, uct_md_h md,
...
-    srq_init_attr.attr.srq_limit = 0;
-    srq_init_attr.srq_context    = self;
-    self->rx.srq.srq = ibv_create_srq(uct_ib_iface_md(&self->super)->pd, &srq_init_attr);
-    if (self->rx.srq.srq == NULL) {
-        ucs_error("failed to create SRQ: %m");
-        status = UCS_ERR_IO_ERROR;
-        goto err_free_tx_ops;
-    }
-    self->rx.srq.available       = srq_init_attr.attr.max_wr;
+    if (srq_size > 0) {
```
Comment (shamisp) [inline src/uct/ib/rc/base/rc_iface.c]: SRQ is optional in IBTA

Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: the current implementation of RC in UCX requires SRQ for scalability

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1161/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3090/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3095/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3096/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3099/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3101/ for details (Mellanox internal link).

**Pull request merged**